x-pack/plugins/timelines/public/components/t_grid/toolbar/fields_browser/index.tsx

```diff
@@ -75,46 +68,58 @@ export const StatefulFieldsBrowserComponent: React.FC<FieldBrowserProps> = ({
+              Object.keys(filteredBrowserFieldsByCategory).length >
+                Object.keys(filteredBrowserFieldsBySelected).length
+              ? category
+              : selected;
+          }, Object.keys(newFilteredBrowserFields)[0]);
+  }, [appliedFilterInput, newFilteredBrowserFields]);
+
   /** Invoked when the user types in the filter input */
-  const updateFilter = useCallback(
-    (newFilterInput: string) => {
-      setFilterInput(newFilterInput);
-      setIsSearching(true);
-      if (inputTimeoutId.current !== 0) {
-        clearTimeout(inputTimeoutId.current); // ⚠️ mutation: cancel any previous timers
-      }
-      // ⚠️ mutation: schedule a new timer that will apply the filter when it fires:
-      inputTimeoutId.current = window.setTimeout(() => {
-        const newFilteredBrowserFields = filterBrowserFieldsByFieldName({
-          browserFields: mergeBrowserFieldsWithDefaultCategory(browserFields),
-          substring: newFilterInput,
-        });
-        setFilteredBrowserFields(newFilteredBrowserFields);
-        setIsSearching(false);
-
-        const newSelectedCategoryId =
-          newFilterInput === '' || Object.keys(newFilteredBrowserFields).length === 0
-            ? DEFAULT_CATEGORY_NAME
-            : Object.keys(newFilteredBrowserFields)
-                .sort()
-                .reduce<string>(
-                  (selected, category) =>
-                    newFilteredBrowserFields[category].fields != null &&
-                    newFilteredBrowserFields[selected].fields != null &&
-                    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-                    Object.keys(newFilteredBrowserFields[category].fields!).length >
-                      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-                      Object.keys(newFilteredBrowserFields[selected].fields!).length
-                      ? category
-                      : selected,
-                  Object.keys(newFilteredBrowserFields)[0]
-                );
-        setSelectedCategoryId(newSelectedCategoryId);
-      }, INPUT_TIMEOUT);
-    },
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    [browserFields, filterInput, inputTimeoutId.current]
-  );
+  const updateFilter = useCallback((newFilterInput: string) => {
+    setFilterInput(newFilterInput);
+    setIsSearching(true);
+  }, []);
+
+  useEffect(() => {
+    if (inputTimeoutId.current !== 0) {
+      clearTimeout(inputTimeoutId.current); // ⚠️ mutation: cancel any previous timers
+    }
+    // ⚠️ mutation: schedule a new timer that will apply the filter when it fires:
+    inputTimeoutId.current = window.setTimeout(() => {
+      setAppliedFilterInput(filterInput);
+    }, INPUT_TIMEOUT);
+    return () => {
+      clearTimeout(inputTimeoutId.current);
+    };
+  }, [filterInput]);
+
+  useEffect(() => {
+    setFilteredBrowserFields(newFilteredBrowserFields);
```

Comment:
Need to test this out, but might there be a benefit to having this in a separate `useEffect()` from the `setSelectedCategoryId` since `newSelectedCategoryId` below is dependent on the `newFilteredBrowserFields`? I think you'll end up with this function running twice versus once if it lived in it's own `useEffect`